common/lib/xmodule/xmodule/modulestore/mongo/base.py

```diff
@@ -193,18 +208,40 @@ def __init__(self, modulestore, course_key, module_data, default_class, cached_m
         # define an attribute here as well, even though it's None
         self.course_id = course_key
         self.cached_metadata = cached_metadata
+        self.cached_modules = {}
+
+    def get_block(self, usage_id):
+        """
+        Overrides the base impl to figure out if you want draft or published and then
+        to strip the revision if it got a draft
+        """
+        if (
+            self.modulestore.get_branch_setting() == ModuleStoreEnum.Branch.draft_preferred
+            and usage_id.category in DIRECT_ONLY_CATEGORIES
+        ):
+            try:
+                block = super(CachingDescriptorSystem, self).get_block(as_draft(usage_id))
+                return wrap_draft(block)
+            except ItemNotFoundError:
+                pass  # do the below
+        block = super(CachingDescriptorSystem, self).get_block(usage_id)
+        return wrap_draft(block)
 
     def load_item(self, location):
         """
         Return an XModule instance for the specified location
         """
         assert isinstance(location, UsageKey)
+        location = self.modulestore.fill_in_run(location)
+        if location in self.cached_modules:
+            return self.cached_modules[location]
         json_data = self.module_data.get(location)
-        if json_data is None:
+        if json_data is None and location not in self.module_data:
```

Comment:
this might read better as just `if location not in self.module_data` and then change L244 to `elif self.module_data[location] is not None:`  and unpack json_data in that block.